File: services/academic-service/src/test/java/com/visor/school/academic/model/TeacherAssignmentTest.java
package com.visor.school.academic.model;

import org.junit.jupiter.api.Test;

import java.util.UUID;

import static org.junit.jupiter.api.Assertions.*;

class TeacherAssignmentTest {

    @Test
    void shouldCreateTeacherAssignment() {
        TeacherAssignment assignment = new TeacherAssignment(
                UUID.randomUUID(),
                UUID.randomUUID(),
                false
        );

        // ID is null until entity is persisted
        assertNull(assignment.getId());
        assertNotNull(assignment.getTeacherId());
        assertNotNull(assignment.getClassId());
        assertFalse(assignment.isClassTeacher());
        assertNotNull(assignment.getAssignedDate());
    }

    @Test
    void shouldCreateClassTeacherAssignmentForGrades7To12() {
        TeacherAssignment assignment = new TeacherAssignment(
                UUID.randomUUID(),
                UUID.randomUUID(),
                true
        );

        assertTrue(assignment.isClassTeacher());
        assertNotNull(assignment.getAssignedDate());
    }

    @Test
    void shouldAcceptAssignedByField() {
        UUID assignedBy = UUID.randomUUID();
        TeacherAssignment assignment = new TeacherAssignment(
                UUID.randomUUID(),
                UUID.randomUUID(),
                false
        );
        assignment.setAssignedBy(assignedBy);

        assertEquals(assignedBy, assignment.getAssignedBy());
    }

    @Test
    void shouldHaveUniqueTeacherAndClassCombination() {
        UUID teacherId = UUID.randomUUID();
        UUID classId = UUID.randomUUID();

        TeacherAssignment assignment1 = new TeacherAssignment(
                teacherId,
                classId,
                false
        );

        TeacherAssignment assignment2 = new TeacherAssignment(
                teacherId,
                classId,
                true // Different isClassTeacher, but same teacher-class
        );

        // Both should be created, but in database would have unique constraint
        // IDs are null until entities are persisted
        assertNull(assignment1.getId());
        assertNull(assignment2.getId());
        assertEquals(teacherId, assignment1.getTeacherId());
        assertEquals(teacherId, assignment2.getTeacherId());
        assertEquals(classId, assignment1.getClassId());
        assertEquals(classId, assignment2.getClassId());
    }
}

File: services/academic-service/src/test/kotlin/com/visor/school/academicservice/model/TeacherAssignmentTest.kt
package com.visor.school.academicservice.model

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import java.util.UUID

class TeacherAssignmentTest {

    @Test
    fun `should create teacher assignment`() {
        val assignment = TeacherAssignment(
            teacherId = UUID.randomUUID(),
            classId = UUID.randomUUID(),
            isClassTeacher = false
        )

        assertNotNull(assignment.id)
        assertNotNull(assignment.teacherId)
        assertNotNull(assignment.classId)
        assertFalse(assignment.isClassTeacher)
        assertNotNull(assignment.assignedDate)
    }

    @Test
    fun `should create class teacher assignment for grades 7-12`() {
        val assignment = TeacherAssignment(
            teacherId = UUID.randomUUID(),
            classId = UUID.randomUUID(),
            isClassTeacher = true
        )

        assertTrue(assignment.isClassTeacher)
        assertNotNull(assignment.assignedDate)
    }

    @Test
    fun `should accept assignedBy field`() {
        val assignedBy = UUID.randomUUID()
        val assignment = TeacherAssignment(
            teacherId = UUID.randomUUID(),
            classId = UUID.randomUUID(),
            isClassTeacher = false,
            assignedBy = assignedBy
        )

        assertEquals(assignedBy, assignment.assignedBy)
    }

    @Test
    fun `should have unique teacher and class combination`() {
        val teacherId = UUID.randomUUID()
        val classId = UUID.randomUUID()

        val assignment1 = TeacherAssignment(
            teacherId = teacherId,
            classId = classId,
            isClassTeacher = false
        )

        val assignment2 = TeacherAssignment(
            teacherId = teacherId,
            classId = classId,
            isClassTeacher = true // Different isClassTeacher, but same teacher-class
        )

        // Both should be created, but in database would have unique constraint
        assertNotNull(assignment1.id)
        assertNotNull(assignment2.id)
        assertEquals(teacherId, assignment1.teacherId)
        assertEquals(teacherId, assignment2.teacherId)
        assertEquals(classId, assignment1.classId)
        assertEquals(classId, assignment2.classId)
    }
}

